builtin/providers/azure/resource_azure_instance.go

```diff
@@ -0,0 +1,653 @@
+				ForceNew: true,
+			},
+
+			"automatic_updates": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"time_zone": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"username": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"password": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"ssh_key_thumbprint": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"endpoint": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+
+						"protocol": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+
+						"public_port": &schema.Schema{
+							Type:     schema.TypeInt,
+							Required: true,
+						},
+
+						"private_port": &schema.Schema{
+							Type:     schema.TypeInt,
+							Required: true,
+						},
+					},
+				},
+				Set: resourceAzureEndpointHash,
+			},
+
+			"security_group": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+
+			"ip_address": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"vip_address": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAzureInstanceCreate(d *schema.ResourceData, meta interface{}) (err error) {
+	mc := meta.(management.Client)
+
+	name := d.Get("name").(string)
+
+	// Compute/set the description
+	description := d.Get("description").(string)
+	if description == "" {
+		description = name
+	}
+
+	// Retrieve the needed details of the image
+	configureForImage, osType, err := retrieveImageDetails(
+		mc,
+		d.Get("image").(string),
+		d.Get("storage").(string),
+	)
+	if err != nil {
+		return err
+	}
+
+	// Verify if we have all required parameters
+	if err := verifyInstanceParameters(d, osType); err != nil {
+		return err
+	}
+
+	p := hostedservice.CreateHostedServiceParameters{
+		ServiceName:    name,
+		Label:          base64.StdEncoding.EncodeToString([]byte(name)),
+		Description:    fmt.Sprintf("Cloud Service created automatically for instance %s", name),
+		Location:       d.Get("location").(string),
+		ReverseDNSFqdn: d.Get("reverse_dns").(string),
+	}
+
+	log.Printf("[DEBUG] Creating Cloud Service for instance: %s", name)
+	err = hostedservice.NewClient(mc).CreateHostedService(p)
+	if err != nil {
+		return fmt.Errorf("Error creating Cloud Service for instance %s: %s", name, err)
+	}
+
+	// Put in this defer here, so we are sure to cleanup already created parts
+	// when we exit with an error
+	defer func(mc management.Client) {
```

Comment:
Wasn't Terraform supposed to not mess with anything if a failure had occurred to facilitate someone coming in and picking off where it had stopped?
If so, the existence of this defer statement is questionable.

In the case of separate hosted service resources; this would have not been necessary anyhow, because they would have had to have been created beforehand...